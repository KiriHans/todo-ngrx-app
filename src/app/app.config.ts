import { ApplicationConfig, isDevMode } from '@angular/core';
import { provideRouter } from '@angular/router';

import { provideAnimations } from '@angular/platform-browser/animations';

import { routes } from './app.routes';
import { provideStore } from '@ngrx/store';
import { provideStoreDevtools } from '@ngrx/store-devtools';
import { appReducers, metaReducers } from './app.reducer';

export const appConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes),
    provideAnimations(),
    provideStore(appReducers, { metaReducers }),
    provideStoreDevtools({ maxAge: 25, logOnly: !isDevMode() }),
  ],
};
