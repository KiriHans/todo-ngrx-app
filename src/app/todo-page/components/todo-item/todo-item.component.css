:host{
	width: 100%;
}

:host ::ng-deep :is(.field-edit-todo) .mat-mdc-form-field-subscript-wrapper {
	display: none;
}

.edit{
  display: none;
}


li{
	list-style: none;
	display: flex;
	justify-content: space-between;
	flex-direction: row;
	align-items: center;
}

.view{
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	gap: 1rem;

	height: 3.2rem;

	width: 100%;
}

.view label{
	display: flex;
	flex-direction: row;
	align-items: center;

	width: 100%;
	height: 2rem;

	font-size: 1rem;

}

.view mat-checkbox {
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	transform: scale(1.2);
}


li.editing .view {
	display: none;
}


li.editing .edit {
	font-family: roboto;
	font-size: 1rem;
	display: block;
	width: calc(100% - 80px);
	height: 3rem;
	padding:  0 16px;
	margin: 0 0 0 40px;

	border: 2px solid #ff4081;

}

.edit:focus{
	outline: none;
	box-shadow: 0 0 1px 1px rgba(255, 64, 127, .5);
	
}

.completed{
	color: #949494;
	text-decoration: line-through;
}

.completed .edit {
	color: #949494;
	text-decoration: line-through;

	border: 2px solid hsla(340, 100%, 63%, 0.732);
	
}

.completed .destroy{
	color: #ff4081;
	font-variation-settings:
  'FILL' 0,
  'wght' 600,
  'GRAD' 0,
  'opsz' 48
}