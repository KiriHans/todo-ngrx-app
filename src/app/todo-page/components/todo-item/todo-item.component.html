<li [ngClass]="{ completed: todo.completed, editing: editing }">
  <div class="view">
    <mat-checkbox [formControl]="completed"> </mat-checkbox>

    <label (dblclick)="editTodo()" for="add">
      {{ todo.text }}
    </label>
    <button (click)="deleteTodo()" mat-icon-button class="destroy">
      <mat-icon>close</mat-icon>
    </button>
  </div>

  <input
    [cdkTrapFocusAutoCapture]="editing"
    [cdkTrapFocus]="editing"
    [formControl]="inputTodo"
    [value]="todo.text"
    (blur)="stopEdit()"
    (keyup.enter)="stopEdit()"
    appSelect
    class="edit"
    type="text"
  />
</li>
