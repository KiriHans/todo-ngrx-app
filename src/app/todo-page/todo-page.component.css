@font-face {
  font-family: Epilogue;
  src: url('src/assets/fonts/Epilogue-VariableFont_wght.ttf');
}

:host{
  --elevated-card-color: rgb(236, 236, 236);
  --elevation-card: 0px 2px 3px -1px rgba(0, 0, 0, .2), 0px 1px 2px 0px rgba(0, 0, 0, .14), 0px 1px 3px 0px rgba(0, 0, 0, .12);
  --max-width-card: 600px;
  --width-pages-background: 7px;

  display: flex;
  flex-direction: column;
  align-items: center;
}


h2{
  color: #ff4081;
  font-weight: 200;
  font-size: 6rem;
  font-family: 'Epilogue', sans-serif;

  text-align: center;
  align-items: center;
  margin: 4rem 0 3rem;
  
}

header{
  margin: 0 auto;  
  width: min(70%, var(--max-width-card));
}

.todo {
  margin: 1rem 0;
  margin-top: 0;
  width: min(70%, var(--max-width-card));
  padding-top: 0;
  position: relative;
  box-sizing: border-box;

  background-color: #fff;

  display: flex;
  flex-direction: column;
  border-color: var(--elevated-card-color);
  box-shadow: var(--elevation-card);
}

.todo::before,  
.todo::after{
  content: "";
  position: absolute;
  top: 50%;
  left: 100%;
  width: var(--width-pages-background);
  height: 97%;
  border-radius: 0 3px 3px 0;
  background-color: rgba(203, 203, 203, 0.843);
  transform: translateY(-50%)
}

.todo::before {
  height: 94%;
  left: calc(100% + var(--width-pages-background));
  background-color: rgb(158, 158, 158)
}

.main{
  padding: 0.7rem;
}

app-footer{
 
  margin-top: 2rem;
  width: min(70%, var(--max-width-card));
}