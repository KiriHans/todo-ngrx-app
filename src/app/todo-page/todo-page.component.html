<section class="todo">
  <h2>Todo</h2>
  <app-todo-add></app-todo-add>

  <app-todo-list> </app-todo-list>
</section>
