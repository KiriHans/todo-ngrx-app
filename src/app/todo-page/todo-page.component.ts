import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TodoAddComponent } from './components/todo-add/todo-add.component';
import { TodoListComponent } from './components/todo-list/todo-list.component';

@Component({
  selector: 'app-todo-page',
  standalone: true,
  imports: [CommonModule, TodoAddComponent, TodoListComponent],
  templateUrl: './todo-page.component.html',
  styleUrls: ['./todo-page.component.css'],
})
export class TodoPageComponent {}
