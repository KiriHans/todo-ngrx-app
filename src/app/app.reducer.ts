import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import { TodoItem } from './todo-page/models/todo-item.model';
import { todoReducer } from './todo-page/store/reducers/todo.reducer';
import { filterReducer } from './todo-page/store/reducers/filter.reducer';
import { ValidFilters } from './todo-page/types/valid-filter.type';
import { storageMetaReducer } from './todo-page/store/reducers/todo.metareducer';

export interface AppState {
  todos: TodoItem[];
  filter: ValidFilters;
}

export const appReducers: ActionReducerMap<AppState> = {
  todos: todoReducer,
  filter: filterReducer,
};

export const metaReducers: MetaReducer[] = [storageMetaReducer];
