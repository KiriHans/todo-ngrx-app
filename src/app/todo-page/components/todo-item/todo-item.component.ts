import { Component, DestroyRef, Input, OnInit, inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import { A11yModule } from '@angular/cdk/a11y';
import { FormControl, ReactiveFormsModule, Validators } from '@angular/forms';
import { SelectInputDirective } from '../../directives/select-input.directive';

/* RXJS Modules */
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';

/* NGRX ModulesÂ¨*/
import { Store } from '@ngrx/store';
import * as TodoActions from '../../store/actions/todo.actions';

/* Angular Material Modules */
import { MatInputModule } from '@angular/material/input';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';

import { TodoItem } from '../../models/todo-item.model';

@Component({
  selector: 'app-todo-item',
  standalone: true,
  imports: [
    CommonModule,
    A11yModule,
    ReactiveFormsModule,
    SelectInputDirective,
    MatInputModule,
    MatCheckboxModule,
    MatButtonModule,
    MatIconModule,
  ],
  templateUrl: './todo-item.component.html',
  styleUrls: ['./todo-item.component.css'],
})
export class TodoItemComponent implements OnInit {
  @Input() todo!: TodoItem;

  private destroyRef = inject(DestroyRef);

  editing = false;
  show = false;

  completed!: FormControl;
  inputTodo!: FormControl;

  constructor(private store: Store) {}

  ngOnInit(): void {
    this.completed = new FormControl(this.todo.completed);
    this.inputTodo = new FormControl(this.todo.text, Validators.required);

    this.completed.valueChanges.pipe(takeUntilDestroyed(this.destroyRef)).subscribe({
      next: () => {
        this.store.dispatch(TodoActions.toggleTodo({ id: this.todo.id }));
      },
    });
  }

  editTodo(): void {
    this.editing = true;
    this.show = true;

    this.inputTodo.setValue(this.todo.text);
  }

  stopEdit(): void {
    this.editing = false;
    this.show = false;

    if (this.inputTodo.invalid || this.inputTodo.value === this.todo.text) return;

    this.store.dispatch(TodoActions.editTodo({ id: this.todo.id, newText: this.inputTodo.value }));
  }

  deleteTodo(): void {
    this.store.dispatch(TodoActions.deleteTodo({ id: this.todo.id }));
  }
}
