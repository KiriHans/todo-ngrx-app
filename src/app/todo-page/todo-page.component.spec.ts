import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TodoPageComponent } from './todo-page.component';
import { RouterTestingModule } from '@angular/router/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { AppState } from '../app.reducer';
import { TodoItem } from './models/todo-item.model';
import { provideMockStore } from '@ngrx/store/testing';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';

describe('TodoPageComponent', () => {
  let component: TodoPageComponent;
  let fixture: ComponentFixture<TodoPageComponent>;
  let el: DebugElement;
  let compiled: HTMLElement;

  const initialState: AppState = {
    todos: [new TodoItem('Test todo'), new TodoItem('Test todo 2')],
    filter: 'all',
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [TodoPageComponent, NoopAnimationsModule, RouterTestingModule],
      providers: [provideMockStore<AppState>({ initialState })],
    });
    fixture = TestBed.createComponent(TodoPageComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should render todo add, list and footer', () => {
    const todoAdd = el.query(By.css('app-todo-add'));
    const todoList = el.query(By.css('app-todo-list'));
    const todoItems = el.queryAll(By.css('app-todo-item'));
    const todoFooter = el.query(By.css('app-todo-footer'));
    const footer = el.query(By.css('app-footer'));

    expect(todoAdd).toBeTruthy();
    expect(todoList).toBeTruthy();
    expect(todoItems).toBeTruthy();
    expect(todoItems).toHaveLength(initialState.todos.length);
    expect(todoFooter).toBeTruthy();
    expect(footer).toBeTruthy();
  });

  it('should have an h2 Element', () => {
    const title = el.query(By.css('h2'));

    expect(title).toBeTruthy();
  });
});
