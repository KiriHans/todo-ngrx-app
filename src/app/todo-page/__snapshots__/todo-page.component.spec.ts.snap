// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoPageComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <header>
    <h2>
      Todo
    </h2>
    <app-todo-add>
      <mat-form-field
        class="mat-mdc-form-field field-add-todo mat-mdc-form-field-type-mat-input mat-mdc-form-field-has-icon-prefix mat-mdc-form-field-has-icon-suffix mat-form-field-no-animations mat-form-field-appearance-fill mat-primary ng-untouched ng-pristine ng-invalid"
      >
        
        <div
          class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled mdc-text-field--no-label"
        >
          <div
            class="mat-mdc-form-field-focus-overlay"
          />
          
          <div
            class="mat-mdc-form-field-flex"
          >
            
            <div
              class="mat-mdc-form-field-icon-prefix"
            >
              <button
                class="mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
                mat-button-ripple-uninitialized=""
                mat-icon-button=""
                maticonprefix=""
              >
                <span
                  class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
                />
                <mat-icon
                  aria-hidden="true"
                  class="mat-icon notranslate toggle-icon no-completed material-icons mat-ligature-font mat-icon-no-color"
                  data-mat-icon-type="font"
                  role="img"
                >
                  <span
                    class="material-symbols-outlined"
                  >
                     toggle_off 
                  </span>
                </mat-icon>
                
                
                <span
                  class="mat-mdc-focus-indicator"
                />
                <span
                  class="mat-mdc-button-touch-target"
                />
              </button>
            </div>
            
            
            <div
              class="mat-mdc-form-field-infix"
            >
              
              
              
              <input
                aria-required="true"
                class="mat-mdc-input-element input-create mat-mdc-form-field-input-control mdc-text-field__input ng-untouched ng-pristine ng-invalid cdk-text-field-autofill-monitored"
                matinput=""
                placeholder="What do you want to do?"
                required=""
              />
            </div>
            
            <div
              class="mat-mdc-form-field-icon-suffix"
            >
              <button
                class="add-button mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
                mat-button-ripple-uninitialized=""
                mat-icon-button=""
                maticonsuffix=""
              >
                <span
                  class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
                />
                <mat-icon
                  aria-hidden="true"
                  class="mat-icon notranslate material-symbols-outlined add-icon material-icons mat-ligature-font mat-icon-no-color"
                  data-mat-icon-type="font"
                  role="img"
                >
                   add_outline 
                </mat-icon>
                <span
                  class="mat-mdc-focus-indicator"
                />
                <span
                  class="mat-mdc-button-touch-target"
                />
              </button>
            </div>
            
          </div>
          <div
            class="mdc-line-ripple mdc-line-ripple--deactivating"
            matformfieldlineripple=""
          />
          
        </div>
        <div
          class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
        >
          
          <div
            class="mat-mdc-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
            style="opacity: 1; transform: translateY(0%);"
          >
            
            <div
              class="mat-mdc-form-field-hint-spacer"
            />
          </div>
          
        </div>
      </mat-form-field>
    </app-todo-add>
  </header>
  <section
    class="todo"
  >
    <section
      class="main"
    >
      <app-todo-list>
        <ul
          class="list"
        >
          <app-todo-item>
            <li>
              <div
                class="view"
              >
                <mat-checkbox
                  class="mat-mdc-checkbox mat-accent _mat-animation-noopable ng-untouched ng-pristine ng-valid"
                >
                  <div
                    class="mdc-form-field"
                  >
                    <div
                      class="mdc-checkbox"
                    >
                      <div
                        class="mat-mdc-checkbox-touch-target"
                      />
                      <input
                        class="mdc-checkbox__native-control"
                        id="mat-mdc-checkbox-3-input"
                        tabindex="0"
                        type="checkbox"
                      />
                      <div
                        class="mdc-checkbox__ripple"
                      />
                      <div
                        class="mdc-checkbox__background"
                      >
                        <svg
                          aria-hidden="true"
                          class="mdc-checkbox__checkmark"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            class="mdc-checkbox__checkmark-path"
                            d="M1.73,12.91 8.1,19.28 22.79,4.59"
                            fill="none"
                          />
                        </svg>
                        <div
                          class="mdc-checkbox__mixedmark"
                        />
                      </div>
                      <div
                        class="mat-ripple mat-mdc-checkbox-ripple mat-mdc-focus-indicator"
                        mat-ripple=""
                      />
                    </div>
                    <label
                      class="mdc-label"
                      for="mat-mdc-checkbox-3-input"
                    />
                  </div>
                </mat-checkbox>
                <label
                  for="edit"
                >
                   Test todo 
                </label>
                <button
                  class="destroy mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
                  mat-button-ripple-uninitialized=""
                  mat-icon-button=""
                >
                  <span
                    class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
                  />
                  <mat-icon
                    aria-hidden="true"
                    class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
                    data-mat-icon-type="font"
                    role="img"
                  >
                    close
                  </mat-icon>
                  <span
                    class="mat-mdc-focus-indicator"
                  />
                  <span
                    class="mat-mdc-button-touch-target"
                  />
                </button>
              </div>
              <div
                aria-hidden="true"
                class="cdk-visually-hidden cdk-focus-trap-anchor"
              />
              <input
                appselect=""
                class="edit ng-untouched ng-pristine ng-valid"
                type="text"
              />
              <div
                aria-hidden="true"
                class="cdk-visually-hidden cdk-focus-trap-anchor"
              />
            </li>
          </app-todo-item>
          <app-todo-item>
            <li>
              <div
                class="view"
              >
                <mat-checkbox
                  class="mat-mdc-checkbox mat-accent _mat-animation-noopable ng-untouched ng-pristine ng-valid"
                >
                  <div
                    class="mdc-form-field"
                  >
                    <div
                      class="mdc-checkbox"
                    >
                      <div
                        class="mat-mdc-checkbox-touch-target"
                      />
                      <input
                        class="mdc-checkbox__native-control"
                        id="mat-mdc-checkbox-4-input"
                        tabindex="0"
                        type="checkbox"
                      />
                      <div
                        class="mdc-checkbox__ripple"
                      />
                      <div
                        class="mdc-checkbox__background"
                      >
                        <svg
                          aria-hidden="true"
                          class="mdc-checkbox__checkmark"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            class="mdc-checkbox__checkmark-path"
                            d="M1.73,12.91 8.1,19.28 22.79,4.59"
                            fill="none"
                          />
                        </svg>
                        <div
                          class="mdc-checkbox__mixedmark"
                        />
                      </div>
                      <div
                        class="mat-ripple mat-mdc-checkbox-ripple mat-mdc-focus-indicator"
                        mat-ripple=""
                      />
                    </div>
                    <label
                      class="mdc-label"
                      for="mat-mdc-checkbox-4-input"
                    />
                  </div>
                </mat-checkbox>
                <label
                  for="edit"
                >
                   Test todo 2 
                </label>
                <button
                  class="destroy mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
                  mat-button-ripple-uninitialized=""
                  mat-icon-button=""
                >
                  <span
                    class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
                  />
                  <mat-icon
                    aria-hidden="true"
                    class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
                    data-mat-icon-type="font"
                    role="img"
                  >
                    close
                  </mat-icon>
                  <span
                    class="mat-mdc-focus-indicator"
                  />
                  <span
                    class="mat-mdc-button-touch-target"
                  />
                </button>
              </div>
              <div
                aria-hidden="true"
                class="cdk-visually-hidden cdk-focus-trap-anchor"
              />
              <input
                appselect=""
                class="edit ng-untouched ng-pristine ng-valid"
                type="text"
              />
              <div
                aria-hidden="true"
                class="cdk-visually-hidden cdk-focus-trap-anchor"
              />
            </li>
          </app-todo-item>
          
        </ul>
        
      </app-todo-list>
    </section>
    <mat-divider
      aria-orientation="horizontal"
      class="mat-divider mat-divider-horizontal"
      role="separator"
    />
    <app-todo-footer>
      <footer>
        <span
          class="todo-count"
        >
          <strong>
            2
          </strong>
           items left 
        </span>
        <ul
          class="filters"
        >
          <li>
            
            <a
              aria-disabled="false"
              class="mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
              href="/#all"
              mat-button=""
              mat-button-ripple-uninitialized=""
              routerlinkactive="active-filter"
            >
              <span
                class="mat-mdc-button-persistent-ripple mdc-button__ripple"
              />
              <span
                class="mdc-button__label"
              >
                 All 
              </span>
              <span
                class="mat-mdc-focus-indicator"
              />
              <span
                class="mat-mdc-button-touch-target"
              />
            </a>
            
          </li>
          <li>
            <a
              aria-disabled="false"
              class="mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
              href="/#completed"
              mat-button=""
              mat-button-ripple-uninitialized=""
            >
              <span
                class="mat-mdc-button-persistent-ripple mdc-button__ripple"
              />
              <span
                class="mdc-button__label"
              >
                 Completed 
              </span>
              <span
                class="mat-mdc-focus-indicator"
              />
              <span
                class="mat-mdc-button-touch-target"
              />
            </a>
            
            
          </li>
          <li>
            <a
              aria-disabled="false"
              class="mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
              href="/#pending"
              mat-button=""
              mat-button-ripple-uninitialized=""
            >
              <span
                class="mat-mdc-button-persistent-ripple mdc-button__ripple"
              />
              <span
                class="mdc-button__label"
              >
                 Pending 
              </span>
              <span
                class="mat-mdc-focus-indicator"
              />
              <span
                class="mat-mdc-button-touch-target"
              />
            </a>
            
            
          </li>
          
        </ul>
        
        <button
          class="clear-completed mdc-button mdc-button--unelevated mat-mdc-unelevated-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
          fragment="all"
          mat-button-ripple-uninitialized=""
          mat-flat-button=""
          tabindex="0"
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-button__ripple"
          />
          <span
            class="mdc-button__label"
          >
             Clear completed 
          </span>
          <span
            class="mat-mdc-focus-indicator"
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </button>
        
      </footer>
    </app-todo-footer>
  </section>
  <app-footer>
    <footer
      class="information"
    >
      <p>
        <strong>
           Double-click 
        </strong>
         to edit a todo
      </p>
      <p>
        <strong>
           Enter 
        </strong>
         to add a new todo
      </p>
      <p>
        Project created using Angular V16
      </p>
    </footer>
  </app-footer>
</div>
`;
